### PR TITLE
[FLINK-17879][python] update Python row serializer to support RowKind

### PR DESCRIPTION
**Pull request by WeiZhong94**
## What is the purpose of the change

*This pull request updates Python row serializer to support RowKind.*

## Brief change log

  - *Support setting and getting row kind from the Row object.*
  - *Support serializing/deserializing the Row/RowData object with RowKind.*

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (yes)
  - The runtime per-record code paths (performance sensitive): (yes)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by WeiZhong94** [comment]
@flinkbot  run azure

File: flink-python/pyflink/fn_execution/coder_impl_fast.pyx [hunk 1]
```diff
@@ -80,7 +81,7 @@ cdef class DataStreamStatelessMapCoderImpl(FlattenRowCoderImpl):
         self._tmp_output_pos = 0
 
     cpdef void _encode_field(self, CoderType coder_type, TypeName field_type, FieldCoder field_coder,
-                        item):
+                             item):
```
Comment (HuangXingBo) [inline flink-python/pyflink/fn_execution/coder_impl_fast.pyx:84]: unnecessary change?

File: flink-python/pyflink/fn_execution/coder_impl_fast.pyx [hunk 2]
```diff
@@ -89,7 +90,7 @@ cdef class DataStreamStatelessMapCoderImpl(FlattenRowCoderImpl):
             self._encode_data_stream_field_complex(field_type, field_coder, item)
 
     cpdef object _decode_field(self, CoderType coder_type, TypeName field_type,
-                        FieldCoder field_coder):
+                               FieldCoder field_coder):
```
Comment (HuangXingBo) [inline flink-python/pyflink/fn_execution/coder_impl_fast.pyx:93]: unnecessary change?

File: flink-python/pyflink/fn_execution/coder_impl_fast.pxd [hunk 3]
```diff
@@ -26,15 +26,17 @@ cdef class BaseCoderImpl:
     cpdef decode_from_stream(self, LengthPrefixInputStream input_stream)
 
 cdef class FlattenRowCoderImpl(BaseCoderImpl):
+    cdef readonly unsigned char ROW_KIND_BIT_SIZE
```
Comment (HuangXingBo) [inline flink-python/pyflink/fn_execution/coder_impl_fast.pxd:29]:
```suggestion
    cdef unsigned char ROW_KIND_BIT_SIZE
```

File: flink-python/pyflink/fn_execution/beam/beam_coder_impl_slow.py [hunk 4]
```diff
@@ -75,62 +82,106 @@ def decode_from_stream(self, in_stream, nested):
...
+            self, in_stream: create_InputStream, nested: bool) -> Tuple[int, List]:
+        row_kind_and_null_mask = self._read_mask(in_stream)
+        row_kind_value = 0
+        for i in range(ROW_KIND_BIT_SIZE):
+            row_kind_value += int(row_kind_and_null_mask[ROW_KIND_BIT_SIZE - i - 1]) * 2 ** i
+        return row_kind_value, [None if row_kind_and_null_mask[idx + ROW_KIND_BIT_SIZE] else
+                                self._field_coders[idx].decode_from_stream(
+                                    in_stream, nested) for idx in range(0, self._field_count)]
+
+    def encode_to_stream(self, iter_value: Iterator[Tuple[int, List]], out_stream, nested):
```
Comment (HuangXingBo) [inline flink-python/pyflink/fn_execution/beam/beam_coder_impl_slow.py:153]: Maybe this method will not be called?

File: flink-python/pyflink/fn_execution/beam/beam_coder_impl_slow.py [hunk 5]
```diff
@@ -75,62 +82,106 @@ def decode_from_stream(self, in_stream, nested):
...
-        return null_mask
+            mask.extend(mask_search_table[b][0:remaining_bits_num])
+        return mask
 
     def __repr__(self):
         return 'FlattenRowCoderImpl[%s]' % ', '.join(str(c) for c in self._field_coders)
 
 
-class RowCoderImpl(FlattenRowCoderImpl):
+class FlattenRowCoderWithRowKindImpl(FlattenRowCoderImpl):
```
Comment (HuangXingBo) [inline flink-python/pyflink/fn_execution/beam/beam_coder_impl_slow.py:138]: Why not we need to introduce `FlattenRowCoderWithRowKindImpl `? What about moving the override method `_decode_one_row_from_stream ` to `RowCoderImpl`

**Review by HuangXingBo — COMMENTED** [review]
@WeiZhong94 Thanks a lot for the PR. I have left some comments.

**Comment by WeiZhong94** [comment]
@HuangXingBo Thanks for your review! I have updated this PR.

**Comment by HuangXingBo** [comment]
@WeiZhong94 Thanks a lot for the update. e2e test of Azure CI failed.

**Comment by WeiZhong94** [comment]
@flinkbot run azure

**Comment by WeiZhong94** [comment]
@flinkbot run azure

File: flink-python/pyflink/fn_execution/beam/beam_coder_impl_slow.py [hunk 6]
```diff
@@ -20,31 +20,38 @@
...
 class FlattenRowCoderImpl(StreamCoderImpl):
 
     def __init__(self, field_coders):
         self._field_coders = field_coders
         self._field_count = len(field_coders)
-        self._leading_complete_bytes_num = self._field_count // 8
-        self._remaining_bits_num = self._field_count % 8
-        self.null_mask_search_table = self.generate_null_mask_search_table()
-        self.null_byte_search_table = (0x80, 0x40, 0x20, 0x10, 0x08, 0x04, 0x02, 0x01)
+        # the row kind uses the first 2 bits of the bitmap, followings are belong to the null mask
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/beam/beam_coder_impl_slow.py:42]:
```suggestion
        # the row kind uses the first 2 bits of the bitmap, the remaining bits are used for null mask
```

File: flink-python/pyflink/datastream/data_stream.py [hunk 7]
```diff
@@ -301,7 +301,8 @@ def key_by(self, key_selector: Union[Callable, KeySelector],
             key_type_info = Types.PICKLED_BYTE_ARRAY()
             is_key_pickled_byte_array = True
 
-        intermediate_map_stream = self.map(lambda x: (key_selector.get_key(x), x),
+        from pyflink.table import Row
```
Comment (dianfu) [inline flink-python/pyflink/datastream/data_stream.py:304]: It seems that we should move Row under module **common**?

File: flink-python/pyflink/fn_execution/beam/beam_coder_impl_slow.py [hunk 8]
```diff
@@ -20,31 +20,38 @@
...
+        # for more details refer to:
+        # https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/RowSerializer.java
+        self._leading_complete_bytes_num = (self._field_count + ROW_KIND_BIT_SIZE) // 8
+        self._remaining_bits_num = (self._field_count + ROW_KIND_BIT_SIZE) % 8
+        self.mask_search_table = self.generate_mask_search_table()
+        self.mask_byte_search_table = (0x80, 0x40, 0x20, 0x10, 0x08, 0x04, 0x02, 0x01)
+        self.row_kind_byte_table = \
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/beam/beam_coder_impl_slow.py:49]:
```suggestion
        self.row_kind_search_table = \
```

File: flink-python/pyflink/fn_execution/beam/beam_coder_impl_slow.py [hunk 9]
```diff
@@ -20,31 +20,38 @@
...
-        self._leading_complete_bytes_num = self._field_count // 8
-        self._remaining_bits_num = self._field_count % 8
-        self.null_mask_search_table = self.generate_null_mask_search_table()
-        self.null_byte_search_table = (0x80, 0x40, 0x20, 0x10, 0x08, 0x04, 0x02, 0x01)
+        # the row kind uses the first 2 bits of the bitmap, followings are belong to the null mask
+        # for more details refer to:
+        # https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/RowSerializer.java
+        self._leading_complete_bytes_num = (self._field_count + ROW_KIND_BIT_SIZE) // 8
+        self._remaining_bits_num = (self._field_count + ROW_KIND_BIT_SIZE) % 8
+        self.mask_search_table = self.generate_mask_search_table()
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/beam/beam_coder_impl_slow.py:47]: I think it should still be **null_mask_search_table**

File: flink-python/pyflink/table/types.py [hunk 10]
```diff
@@ -2068,6 +2079,12 @@ def conv(obj):
         else:
             return dict(zip(self._fields, self))
 
+    def get_row_kind(self) -> RowKind:
```
Comment (dianfu) [inline flink-python/pyflink/table/types.py:2082]: What about changed as following?
```suggestion
    @property
    def row_kind(self) -> RowKind:
```

File: flink-python/pyflink/table/types.py [hunk 11]
```diff
@@ -2068,6 +2079,12 @@ def conv(obj):
...
+        return self._row_kind
+
+    def set_row_kind(self, row_kind: RowKind):
```
Comment (dianfu) [inline flink-python/pyflink/table/types.py:2085]:
```suggestion
    @row_kind.setter
    def row_kind(self, row_kind: RowKind):
```

File: flink-python/pyflink/fn_execution/beam/beam_coder_impl_slow.py [hunk 12]
```diff
@@ -121,16 +140,29 @@ class RowCoderImpl(FlattenRowCoderImpl):
...
                 field_coders[i].encode_to_stream(item, out_stream, nested)
 
     def decode_from_stream(self, in_stream, nested):
-        return Row(*self._decode_one_row_from_stream(in_stream, nested))
+        row_kind_value, fields = self._decode_one_row_from_stream(in_stream, nested)
+        row = Row(*fields)
+        row.set_row_kind(RowKind(row_kind_value))
+        return row
+
+    def _decode_one_row_from_stream(
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/beam/beam_coder_impl_slow.py:157]: Could we improve the implementation in the super class?

File: flink-python/pyflink/table/types.py (hunk 10 above)
Comment (WeiZhong94) [reply flink-python/pyflink/table/types.py]: The `Row` class has override the "__getattr__" method to access fields via column names. If we change to this, it will cause an ambiguity problem.

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your review! I have addressed your comments.

**Review by dianfu — APPROVED** [review]
